circuit/environment/src/helpers/linear_combination.rs

```diff
@@ -463,6 +459,17 @@ impl<F: PrimeField> Mul<&F> for &LinearCombination<F> {
     }
 }
 
+impl<F: PrimeField> MulAssign<&F> for LinearCombination<F> {
+    fn mul_assign(&mut self, coefficient: &F) {
+        self.constant *= coefficient;
+        self.terms.iter_mut().for_each(|(_v, current_coefficient)| {
+            *current_coefficient *= coefficient;
+        });
+        self.terms.retain(|(_v, res)| !res.is_zero());
+        self.value *= coefficient;
+    }
+}
+
```

Comment:
This struct in the `circuit` crate is a central object in constructing the R1CS so there's zero room for error here.  

At the very least, there needs to be several unit tests written around this operation to ensure the correct amount of variables and constraints come out of this. Can you write unit tests surrounding it both here and in the mul operation that's employing this change? 

It should also be ensured any tests that employ test vectors which explicitly or implicitly employ the `mul` operation are re-run (if they're not done so by default in CI).

Reply:
While I believe that this change (`filter_map` changed to `for_each`+`retain`) doesn't result in any changes to crypto logic (i.e. the order and number of operations involved), I agree that it should be reviewed carefully :+1:.

Since I'm not a crypto expert, I will let the other team members decide if the existing tests are sufficient or if we need more of them (and how they would need to look like).

In the meantime, I'll be happy to add a relevant benchmark.

Reply:
@iamalwaysuncomfortable I've [updated](https://github.com/AleoNet/snarkVM/pull/2460/commits/79f2a1309ed0a9d5e1122b2c4c0be4d912459ae6) the PR so that it doesn't affect the existing `Mul` impl, and uses a virtually identical implementation :+1:.